### PR TITLE
kexec: add support for adding more ACPI tables

### PR DESCRIPTION
**Pull request by rminnich**
In some cases we want to extend the ACPI tables with more.

This change adds a -a switch to kexec, which allows one to add
a single table. For now that is all we need; if we want more we
change the operation later, once we are sure this is the usage
mode we need.

Note that due to Linux limitations, the -a switch only
works with multiboot kernels.

This is tested and seems to work with our toy multiboot kernel.

### CODE REVIEW

File: cmds/kexec/kexec_linux.go [hunk 1]
```diff
@@ -135,21 +141,72 @@ func main() {
...
+			log.Fatal(err)
+		}
+	}
+
+	var segs []kexec.Segment // used to add other segments for, e.g., multiboot
+	if opts.acpi != "" {
+		if !mbk {
+			log.Fatalf("You can only specify -a with multiboot kernels")
+		}
+		// NewSDT won't work on some linux kernels that limit reading
```
Comment (rjoleary) [inline cmds/kexec/kexec_linux.go]: The inside of this if-statement could be moved to a function, maybe called `loadACPISegmentForMultiboot`.

File: cmds/kexec/kexec_linux.go [hunk 2]
```diff
@@ -135,21 +141,72 @@ func main() {
...
+			log.Fatal(err)
+		}
+		rr = append(rr, xtra)
+		log.Printf("Calling SDT Marshal with %d tables", rr)
+		b, err := s.MarshalAll(rr...)
+		if err != nil {
+			log.Fatal(err)
+		}
+		log.Printf("Marshal'ed %d bytes for ACPI segment", len(b))
+		seg := kexec.NewSegment(b, kexec.Range{Start: uintptr(s.Base), Size: uint(len(b))})
```
Comment (rjoleary) [inline cmds/kexec/kexec_linux.go]: When you overwrite the ACPI, how do you know there's enough space to accommodate the larger size of `b`? Do you have to update the size of the ACPI range in e820 or anything like that?

File: cmds/kexec/kexec_linux.go [hunk 3]
```diff
@@ -38,11 +39,13 @@ type options struct {
 	load         bool
 	exec         bool
 	debug        bool
+	acpi         string
 	modules      []string
 }
 
 func registerFlags() *options {
 	o := &options{}
+	flag.StringVarP(&o.acpi, "acpi", "a", "", "Add an acpi table")
```
[the remark this answers is not in the document]
Comment (rminnich) [reply cmds/kexec/kexec_linux.go]: done

File: cmds/kexec/kexec_linux.go [hunk 4]
```diff
@@ -120,7 +126,7 @@ func main() {
 		log.Fatalf("--reuse-cmdline and other command line options are mutually exclusive")
 	}
 
-	if opts.load == false && opts.exec == false {
+	if opts.load == false && opts.exec == false && opts.acpi == "" {
```
[the remark this answers is not in the document]
Comment (rminnich) [reply cmds/kexec/kexec_linux.go]: This gets caught in a rather indirect way. If opts.load is not set, the probe of mbk won't get done, so we'll get an error. That may not be a nice way to indicate the problem, however. I improved the error message.

File: cmds/kexec/kexec_linux.go [hunk 5]
```diff
@@ -135,21 +141,72 @@ func main() {
 		}
 	}
 
+	// mbk indicates that we are a multiboot kernel
+	var mbk bool
+	var kernelpath string
 	if opts.load {
-		kernelpath := flag.Args()[0]
+		kernelpath = flag.Args()[0]
```
[the remark this answers is not in the document]
Comment (rminnich) [reply cmds/kexec/kexec_linux.go]: done

File: cmds/kexec/kexec_linux.go [hunk 6]
```diff
@@ -135,21 +141,72 @@ func main() {
...
+			mbk = true
+		} else if err == multiboot.ErrFlagsNotSupported {
+			log.Fatal(err)
+		}
+	}
+
+	var segs []kexec.Segment // used to add other segments for, e.g., multiboot
+	if opts.acpi != "" {
+		if !mbk {
+			log.Fatalf("You can only specify -a with multiboot kernels")
```
[the remark this answers is not in the document]
Comment (rminnich) [reply cmds/kexec/kexec_linux.go]: done

File: cmds/kexec/kexec_linux.go (hunk 1 above)
Comment (rminnich) [reply cmds/kexec/kexec_linux.go]: done

File: cmds/kexec/kexec_linux.go (hunk 2 above)
Comment (rminnich) [reply cmds/kexec/kexec_linux.go]: I'm gonna mark the segment now. This might even work.

File: cmds/kexec/kexec_linux.go [hunk 7]
```diff
@@ -11,6 +11,7 @@
 //		 Loads a kernel for later execution.
 //
 // Options:
+//     -acpi=STRING or -a=string      Add an ACPI table (only one at present)
```
Comment (rjoleary) [inline cmds/kexec/kexec_linux.go]:
```suggestion
//     --acpi=STRING or -a=string     Add an ACPI table (only one at present)
```

File: pkg/multiboot/multiboot_test.go [hunk 8]
```diff
@@ -118,3 +120,38 @@ func TestParseHeader(t *testing.T) {
 		})
 	}
 }
+
+func TestEnsure(t *testing.T) {
+	v = t.Logf
+	v("hi")
```
Comment (rjoleary) [inline pkg/multiboot/multiboot_test.go]: hi

Comment (rminnich) [reply pkg/multiboot/multiboot_test.go]: well hi to you too

this got fixed
:-)

**Pull request merged**